File: example/feature_b/src/main/java/me/vponomarenko/feature/b/FragmentB.java
package me.vponomarenko.feature.b;

import android.os.Bundle;
import android.text.Editable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import me.vponomarenko.core.SimpleTextWatcher;
import me.vponomarenko.core.TextHolder;
import me.vponomarenko.feature.b.di.FeatureBComponent;
import me.vponomarenko.feature.b.di.TextHolderForFeatureB;
import me.vponomarenko.injectionmanager.IHasComponent;
import me.vponomarenko.injectionmanager.x.XInjectionManager;

/**
 * Author: Valery Ponomarenko
 * Date: 25/08/2018
 * LinkedIn: https://www.linkedin.com/in/ponomarenkovalery
 */

public class FragmentB extends Fragment implements IHasComponent<FeatureBComponent> {

    @Inject
    TextHolder singletonTextHolder;

    @Inject
    @TextHolderForFeatureB
    TextHolder featureBTextHolder;

    private SimpleTextWatcher editTextWatcherForSingleton = new SimpleTextWatcher() {
        @Override
        public void afterTextChanged(@NotNull Editable s) {
            if (s.length() == 0) return;
            singletonTextHolder.setText(s.toString());
        }
    };

    private SimpleTextWatcher editTextWatcherForFeatureB = new SimpleTextWatcher() {
        @Override
        public void afterTextChanged(@NotNull Editable s) {
            if (s.length() == 0) return;
            featureBTextHolder.setText(s.toString());
        }
    };

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_feature_b, container, false);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        FeatureBComponent component = XInjectionManager.bindComponent(this);
        component.inject(this);
        if (savedInstanceState == null) {
            getChildFragmentManager().beginTransaction()
                    .replace(R.id.childFragmentContainer, new FragmentChildB())
                    .commit();
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        EditText singleton = view.findViewById(R.id.editText_singleton);
        EditText featureB = view.findViewById(R.id.editText_feature_b);
        singleton.addTextChangedListener(editTextWatcherForSingleton);
        featureB.addTextChangedListener(editTextWatcherForFeatureB);
    }

    @NotNull
    @Override
    public FeatureBComponent getComponent() {
        return FeatureBComponent.Initializer.Companion.init();
    }

    @NotNull
    @Override
    public String getComponentKey() {
        return getClass().toString();
    }
}

File: feature_b/src/main/java/me/vponomarenko/feature/b/FragmentB.kt
package me.vponomarenko.feature.b

import android.os.Bundle
import android.text.Editable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import kotlinx.android.synthetic.main.fragment_feature_b.*
import me.vponomarenko.core.SimpleTextWatcher
import me.vponomarenko.core.TextHolder
import me.vponomarenko.feature.b.di.FeatureBComponent
import me.vponomarenko.feature.b.di.TextHolderForFeatureB
import me.vponomarenko.injectionmanager.IHasComponent
import me.vponomarenko.injectionmanager.InjectionManager
import javax.inject.Inject

/**
 * Author: Valery Ponomarenko
 * Date: 24/08/2018
 * LinkedIn: https://www.linkedin.com/in/ponomarenkovalery
 */

class FragmentB : Fragment(), IHasComponent {

    @Inject
    lateinit var singletonTextHolder: TextHolder

    @Inject
    @field:TextHolderForFeatureB
    lateinit var featureBTextHolder: TextHolder

    private val editTextWatcherForSingleton = object : SimpleTextWatcher() {
        override fun afterTextChanged(s: Editable) {
            if (s.isEmpty()) return
            singletonTextHolder.text = s.toString()
        }
    }

    private val editTextWatcherForFeatureB = object : SimpleTextWatcher() {
        override fun afterTextChanged(s: Editable) {
            if (s.isEmpty()) return
            featureBTextHolder.text = s.toString()
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = inflater.inflate(R.layout.fragment_feature_b, container, false)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        InjectionManager.instance
            .bindComponent<FeatureBComponent>(this)
            .inject(this)
        if (savedInstanceState == null) {
            childFragmentManager.beginTransaction()
                .replace(R.id.childFragmentContainer, FragmentChildB())
                .commit()
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        editText_singleton.addTextChangedListener(editTextWatcherForSingleton)
        editText_feature_b.addTextChangedListener(editTextWatcherForFeatureB)
    }

    override fun createComponent() = FeatureBComponent.Initializer.init()
}